File: 0049-group-anagrams/solution.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        if (!strs.size()) return {{}};

        vector<vector<string>> res;

        // Now create a hashmap to store all elements.
        unordered_map<string, vector<string>> mp;

        // Sort all string elements of input array.
        // Then check if exists in hashmap.
        for (auto& i : strs) {
            string key = i;
            sort(key.begin(), key.end());

            // store all key pair values in hashmap according to sorted key
            mp[key].push_back(i);
        }

        // prepare result
        for (auto& i: mp) {
            res.push_back(i.second);
        }
        return res;
    }
};

File: 0049-group-anagrams/solution.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # we will create a hashmap of Counter : anagrams list []
        # create a counter for each word, and group together the anagrams in that specific list/tuple
        res = defaultdict(list)

        # iterate for every word in the list
        for s in strs:
            # for this specific string, create a counter object
            count = [0] * 26

            # now calculate the character count
            for c in s:
                count[ord(c) - ord("a")] += 1
            
            # map the current string to the count tuple
            res[tuple(count)].append(s)
        
        # now we have a res hashmap of values of multiple lists
        # which are anagrams, we dont need keys (counter) anymore

        return list(res.values())
